README.md
[![PyPI version](https://badge.fury.io/py/dvg-debug-functions.svg?kill_cache=4)](https://badge.fury.io/py/dvg-debug-functions)
[![Build Status](https://travis-ci.org/Dennis-van-Gils/python-dvg-debug-functions.svg?branch=master&kill_cache=4)](https://travis-ci.org/Dennis-van-Gils/python-dvg-debug-functions)
[![Coverage Status](https://coveralls.io/repos/github/Dennis-van-Gils/python-dvg-debug-functions/badge.svg?branch=master&kill_cache=4)](https://coveralls.io/github/Dennis-van-Gils/python-dvg-debug-functions?branch=master)

# DvG_debug_functions
Provides functions for printing debug information to the terminal output.

# dprint()
```Python
dprint(str_msg, ANSI_color=None) 
````
'Debug' print a single line to the terminal with optional ANSI color codes. The line will be terminated with a newline character and the terminal output buffer is forced to flush before and after every print. In addition, if PyQt5 is present in the Python environment, then a mutex lock will be obtained and released again for each dprint execution.

There is a lot of overhead using this print statement, but it is particularly well-suited for multithreaded PyQt programs where multiple threads are each printing information to the same terminal. The `dprint` function ensures that each line sent to the terminal will remain as a continious single line, whereas a regular `print` statement will likely result in the lines getting mixed up.

# tprint()
```Python
tprint(str_msg, ANSI_color=None) 
````
Identical to `dprint(...)`, but now prepended with a `time.perf_counter()` timestamp.

# print_fancy_traceback()
```Python
print_fancy_traceback(err, back=3)
```
Prints the exception `err` to the terminal with a traceback that is `back` deep, using ANSI color codes that mimic the IPython command shell.

Example output:

![print_fancy_traceback.png](https://raw.githubusercontent.com/Dennis-van-Gils/python-dvg-debug-functions/master/images/print_fancy_traceback.png)



CHANGELOG.md
Changelog
=========

0.0.1 (2020-05-13)
------------------
* First release on PyPI.

1.1.0 (2020-06-04)
------------------
* Added `tprint(...)`

1.1.1 (2020-06-06)
------------------
* Extended README

1.1.2 (2020-06-08)
------------------
* Code style: black

tests/test_DvG_debug_functions.py
from DvG_debug_functions import (
    dprint,
    tprint,
    ANSI,
    print_fancy_traceback as pft,
)
from unittest import mock
import io


def test_dprint():
    with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
        dprint("No color")

    assert fake_stdout.getvalue() == "No color\n"


def test_dprint_in_red():
    with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
        dprint("In red", ANSI.RED)

    assert fake_stdout.getvalue() == "\x1b[1;31mIn red\x1b[1;37m\n"


def test_tprint():
    with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
        tprint("No color")

    assert fake_stdout.getvalue()[-9:] == "No color\n"


def test_pft():
    try:
        0 / 0
    except ZeroDivisionError as err:
        with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
            pft(err, 1)

        assert (
            fake_stdout.getvalue().split("\n")[-2]
            == "\x1b[1;31mZeroDivisionError: \x1b[1;37mdivision by zero"
        )


def test_pft_overshoot_callstack():
    try:
        0 / 0
    except ZeroDivisionError as err:
        with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
            pft(err, 50)

        assert (
            fake_stdout.getvalue().split("\n")[-2]
            == "\x1b[1;31mZeroDivisionError: \x1b[1;37mdivision by zero"
        )


def test_pft_err_as_string():
    with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
        pft("Custom error string", 1)

    assert (
        fake_stdout.getvalue().split("\n")[-2]
        == "\x1b[1;31mError: \x1b[1;37mCustom error string"
    )


def test_pft():
    try:
        0 / 0
    except ZeroDivisionError as err:
        with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
            pft(err, 1)


def test_pft_descr_abbr():
    try:
        0 / 0
    except ZeroDivisionError as err:
        err.abbreviation = "abbr"
        err.description = "descr"
        with mock.patch("sys.stdout", new=io.StringIO()) as fake_stdout:
            pft(err, 1)

        assert (
            fake_stdout.getvalue().split("\n")[-2]
            == "\x1b[1;31mZeroDivisionError: \x1b[1;37mabbr: descr"
        )


src/DvG_debug_functions.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Provides functions for neatly printing debug information to the terminal
output, especially well-suited for multithreaded programs.

Functions:
    dprint(...):
        'Debug' print a single line to the terminal with optional ANSI color
        codes. Particularly well-suited for multithreaded PyQt programs where
        multiple threads are printing information to the same terminal.
        
    tprint(...):
        Identical to dprint(...), but now prepended with a time.perf_counter()
        timestamp.

    print_fancy_traceback(...):
        Prints the exception to the terminal, using ANSI color codes that mimic
        the IPython command shell.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/python-dvg-debug-functions"
__date__ = "08-06-2020"
__version__ = "1.1.2"

import os
import sys
import time

try:
    from PyQt5 import QtCore
except:
    PYQT5_IS_PRESENT = False
else:
    PYQT5_IS_PRESENT = True
    dprint_mutex = QtCore.QMutex()


class ANSI:
    NONE = ""
    RED = "\033[1;31m"
    GREEN = "\033[1;32m"
    YELLOW = "\033[1;33m"
    BLUE = "\033[1;34m"
    PURPLE = "\033[1;35m"
    CYAN = "\033[1;36m"
    WHITE = "\033[1;37m"


def dprint(str_msg, ANSI_color=None):
    """'Debug' print a single line to the terminal with optional ANSI color
    codes. The line will be terminated with a newline character and the
    terminal output buffer is forced to flush before and after every print.
    In addition, if PyQt5 is present in the Python environment, then a mutex
    lock will be obtained and released again for each dprint execution.

    There is a lot of overhead using this print statement, but it is
    particularly well-suited for multithreaded PyQt programs where multiple
    threads are each printing information to the same terminal. The dprint
    function ensure that each line sent to the terminal will remain as a
    continious single line, whereas a regular print statement will likely
    result in the lines getting mixed up.
    """
    # Explicitly ending the string with a newline '\n' character, instead
    # of letting the print statement end it for you (end='\n'), fixes the
    # problem of single lines getting printed to the terminal with
    # intermittently delayed newlines when coming from different threads.
    # I.e. it prevents:
    # >: Output line of thread 1Output line of thread 2   (\n)
    # >:                                                  (\n)
    # and makes sure we get:
    # >: Output line of thread 1                          (\n)
    # >: Output line of thread 2                          (\n)

    if PYQT5_IS_PRESENT:
        locker = QtCore.QMutexLocker(dprint_mutex)

    sys.stdout.flush()
    if ANSI_color is None:
        print("%s\n" % str_msg, end="")
    else:
        print("%s%s%s\n" % (ANSI_color, str_msg, ANSI.WHITE), end="")
    sys.stdout.flush()

    if PYQT5_IS_PRESENT:
        locker.unlock()


def tprint(str_msg, ANSI_color=None):
    """Identical to dprint(...), but now prepended with a time.perf_counter()
    timestamp
    """
    dprint("%.4f %s" % (time.perf_counter(), str_msg), ANSI_color)


def print_fancy_traceback(err, back=3):
    """Print the exception `err` to the terminal with a traceback that is
    `back` deep, using ANSI color codes that mimic the IPython command shell.
    """
    print(
        ANSI.WHITE
        + "\nFancy traceback "
        + ANSI.CYAN
        + "(most recent call last)"
        + ANSI.WHITE
        + ":"
    )

    while back >= 1:
        try:
            err_file = os.path.basename(sys._getframe(back).f_code.co_filename)
        except ValueError:
            # Call stack is not deep enough. Proceed to next in line.
            back -= 1
        else:
            err_fun = sys._getframe(back).f_code.co_name
            err_line = sys._getframe(back).f_lineno
            print(
                (
                    ANSI.CYAN
                    + "File "
                    + ANSI.GREEN
                    + '"%s"'
                    + ANSI.CYAN
                    + ", line "
                    + ANSI.GREEN
                    + "%s"
                    + ANSI.CYAN
                    + ", in "
                    + ANSI.PURPLE
                    + "%s"
                    + ANSI.WHITE
                )
                % (err_file, err_line, err_fun)
            )
            back -= 1

    if isinstance(err, Exception):
        if not hasattr(err, "abbreviation"):
            err.abbreviation = ""
        if not hasattr(err, "description"):
            err.description = ""

        if err.abbreviation == "" and err.description == "":
            print(
                (ANSI.RED + "%s: " + ANSI.WHITE) % sys.exc_info()[0].__name__,
                end="",
            )
            print(err)
        else:
            print(
                (ANSI.RED + "%s: " + ANSI.WHITE + "%s: %s")
                % (
                    sys.exc_info()[0].__name__,
                    err.abbreviation,
                    err.description,
                )
            )

    elif isinstance(err, str):
        print((ANSI.RED + "Error: " + ANSI.WHITE + "%s") % err)
